src/LaunchDarkly.EventSource/AsyncHelpers.cs

```diff
@@ -0,0 +1,91 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace LaunchDarkly.EventSource
+{
+    internal static class AsyncHelpers
+    {
+        // General-purpose timeout logic that uses a timed CancellationToken instead of
+        // Task.Delay. The cancellationToken that is passed in is for explicitly cancelling
+        // the operation from elsewhere; DoWithTimeout wraps this to create a token that can
+        // also be cancelled by a timeout, and passes that token to taskFn.
+        internal static async Task<T> DoWithTimeout<T>(
+            TimeSpan timeout,
+            CancellationToken cancellationToken,
+            Func<CancellationToken, Task<T>> taskFn
+            )
+        {
+            var timeoutCancellation = new CancellationTokenSource(timeout);
+            var combinedCancellation = CancellationTokenSource.CreateLinkedTokenSource(
+                cancellationToken, timeoutCancellation.Token);
+            Task<T> task = null;
+            try
+            {
+                task = taskFn(combinedCancellation.Token);
+                return await task;
+            }
+            catch (OperationCanceledException)
+            {
+                SuppressExceptions(task);
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    throw;
+                }
+                throw new ReadTimeoutException();
+            }
+        }
+
+        // Take a task that can't be terminated early with a CancellationToken, and wrap it in a task
+        // that can. If you cancel the wrapped task, the inner task will still continue, but we're
+        // assuming that in these situations we'll eventually be closing the underlying socket which
+        // will terminate any pending I/O.
```

Comment:
Here's the other thing we need for our new improved timeout logic. While the idea of the cancellation token is very nice, and it works well for things like HTTP requests, some of the basic .NET I/O methods either don't have a cancellation token parameter (like `StreamReader.ReadAsync`), or do have one but it doesn't actually work (like `Stream.ReadAsync`— that is, the base class `Stream` declares a method with that parameter, but some of the stream subclasses don't implement it, and unfortunately the one for reading a socket is one of those).

So, we have to use the Task API in a somewhat confusing way to make it so we really can interrupt these operations... or at least stop waiting on them. See code comments below.

Reply:
And I think this is still better than our earlier logic that used `WhenAny`, because that was always creating two _simultaneously active_ tasks every time we wanted to do a read. Here, using `ContinueWith` does create additional tasks, but they are chained— `cancellableTask` isn't taking up any room in the scheduler until `originalTask` finishes.